src/stream/src/executor/row_merge.rs

```diff
@@ -0,0 +1,189 @@
+    pub lhs_input: Executor,
+    pub rhs_input: Executor,
+    /// Maps input from the lhs to the output.
+    pub lhs_mapping: ColIndexMapping,
+    /// Maps input from the rhs to the output.
+    pub rhs_mapping: ColIndexMapping,
+    /// Output schema
+    pub schema: Schema,
+}
+
+impl RowMergeExecutor {
+    pub fn new(
+        ctx: ActorContextRef,
+        lhs_input: Executor,
+        rhs_input: Executor,
+        lhs_mapping: ColIndexMapping,
+        rhs_mapping: ColIndexMapping,
+        schema: Schema,
+    ) -> Self {
+        Self {
+            ctx,
+            lhs_input,
+            rhs_input,
+            lhs_mapping,
+            rhs_mapping,
+            schema,
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    pub async fn execute_inner(self) {
+        let lhs_mapping = self.lhs_mapping;
+        let rhs_mapping = self.rhs_mapping;
+        let data_types = self
+            .schema
+            .fields()
+            .iter()
+            .map(|f| f.data_type())
+            .collect::<Vec<_>>();
+
+        {
+            let mut lhs_buffer: Vec<StreamChunk> = Vec::with_capacity(1);
+            let mut rhs_buffer: Vec<StreamChunk> = Vec::with_capacity(1);
+            let aligned_stream = barrier_align(
+                self.lhs_input.execute(),
+                self.rhs_input.execute(),
+                self.ctx.id,
+                self.ctx.fragment_id,
+                self.ctx.streaming_metrics.clone(),
+                "Join",
+            );
+            pin_mut!(aligned_stream);
+            #[for_await]
+            for message in aligned_stream {
+                match message? {
+                    AlignedMessage::Left(chunk) => {
+                        lhs_buffer.push(chunk);
+                    }
+                    AlignedMessage::Right(chunk) => {
+                        rhs_buffer.push(chunk);
+                    }
+                    AlignedMessage::Barrier(barrier) => {
+                        if lhs_buffer.is_empty() && rhs_buffer.is_empty() {
+                            yield Message::Barrier(barrier);
+                            continue;
+                        }
+                        #[for_await]
+                        for output in Self::flush_buffers(
+                            &data_types,
+                            &lhs_mapping,
+                            &rhs_mapping,
+                            &mut lhs_buffer,
+                            &mut rhs_buffer,
+                        ) {
+                            yield output?;
+                        }
+                        yield Message::Barrier(barrier);
+                    }
+                    AlignedMessage::WatermarkLeft(watermark) => {
+                        yield Message::Watermark(watermark);
+                    }
+                    AlignedMessage::WatermarkRight(watermark) => {
+                        yield Message::Watermark(watermark);
+                    }
```

Comment:
Since we buffer the chunks, we can't emit watermark messages bypassing the chunk; otherwise, the watermark guarantee cannot be maintained. It seems necessary to buffer the watermark messages as well. Refer to [this PR](https://github.com/risingwavelabs/risingwave/pull/17968) for a solution to this issue.

Reply:
I don't think we should see watermark at all actually. Agg should not propagate watermark. I suggest just absorbing and ignoring any watermark, leaving a trace::warn if we see unexpected watermarks.

Reply:
Handled: https://github.com/risingwavelabs/risingwave/pull/17930/commits/ebf0545b60143632ff9e9cc574f3ef1f6e4047df